File: backend/app/schemas/learning.py
"""Request and response contracts for the adaptive learning engine.

The most important thing in this file is the split between `QuizQuestionRead` and
`QuizQuestionResult`. The first is what the student receives while taking a quiz and
deliberately omits `correct_index`, `explanation` and the source; the second is
returned only after an answer is submitted. Answers cannot leak early because the
shape that would carry them is never used on the taking path.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.models import (
    OPTIONS_PER_QUESTION,
    AnswerVerdict,
    Difficulty,
    GradingState,
    QuestionType,
    QuizMode,
)
from app.schemas.common import ORMModel
from app.services.learning.quiz_service import QuizFormat

MAX_QUIZ_QUESTIONS = 20
MIN_QUIZ_QUESTIONS = 3

# Mirrors MAX_ANSWER_CHARS in services/learning/grading.py. Rejected here with a
# clear message rather than silently truncated at the grader.
MAX_RESPONSE_CHARS = 2_000


# --- Topics --------------------------------------------------------------------


class TopicRead(ORMModel):
    id: uuid.UUID
    course_id: uuid.UUID
    name: str
    description: str
    is_active: bool
    created_at: datetime


class TopicExtractionResponse(BaseModel):
    """What changed, so the UI can explain a regeneration rather than just refresh."""

    created: list[TopicRead]
    reactivated: list[TopicRead]
    deactivated: list[TopicRead]
    unchanged: list[TopicRead]


# --- Provenance ----------------------------------------------------------------


class SourceRef(BaseModel):
    """Where a generated item came from, read from the stored chunk's own row."""

    document_id: uuid.UUID
    document_name: str
    # None for formats without real pages (TXT, Markdown) — never a fabricated 1.
    page_number: int | None
    chunk_id: uuid.UUID


# --- Quizzes -------------------------------------------------------------------


class QuizGenerateRequest(BaseModel):
    mode: QuizMode = QuizMode.ADAPTIVE
    question_count: int = Field(default=8, ge=MIN_QUIZ_QUESTIONS, le=MAX_QUIZ_QUESTIONS)
    # Standard mode only. Ignored for adaptive, where the engine chooses.
    topic_ids: list[uuid.UUID] = Field(default_factory=list)
    difficulty: Difficulty | None = None
    # Defaults to multiple choice, so a client written before short answers
    # existed keeps getting exactly the quiz it used to get.
    quiz_format: QuizFormat = QuizFormat.MCQ

    @field_validator("topic_ids")
    @classmethod
    def _cap_topics(cls, value: list[uuid.UUID]) -> list[uuid.UUID]:
        if len(value) > 10:
            raise ValueError("Select at most 10 topics.")
        return value


class QuizQuestionRead(ORMModel):
    """The taking view. No correct answer, no explanation, no source.

    For a short-answer question that also means no `reference_answer`, no
    `key_concepts` and no `rubric`: those are the short-answer equivalent of
    `correct_index`, and shipping them to the client before submission would hand
    the student the mark scheme. The taking path never uses a shape that carries
    them.
    """

    id: uuid.UUID
    position: int
    question_text: str
    question_type: QuestionType
    # Null for short answers — there is nothing to choose from.
    options: list[str] | None
    difficulty: Difficulty
    topic_id: uuid.UUID
    topic_name: str


class QuizRead(ORMModel):
    id: uuid.UUID
    course_id: uuid.UUID
    title: str
    mode: QuizMode
    selection_rationale: str
    difficulty_plan: dict[str, int]
    question_count: int
    created_at: datetime


class QuizDetail(QuizRead):
    questions: list[QuizQuestionRead]


# --- Attempts ------------------------------------------------------------------


class AnswerSubmit(BaseModel):
    question_id: uuid.UUID
    selected_index: int = Field(ge=0, le=OPTIONS_PER_QUESTION - 1)
    answered_in_seconds: int | None = Field(default=None, ge=0, le=86_400)


class ShortAnswerSubmit(BaseModel):
    """A written answer. Separate from `AnswerSubmit` so neither contract has an
    optional field that is really required for one of the two question types."""

    question_id: uuid.UUID
    response_text: str = Field(min_length=1, max_length=MAX_RESPONSE_CHARS)
    answered_in_seconds: int | None = Field(default=None, ge=0, le=86_400)

    @field_validator("response_text")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Write an answer before submitting it.")
        return value


class ConceptResultRead(BaseModel):
    """One rubric line, and whether the answer satisfied it."""

    concept: str
    satisfied: bool


class AnswerResult(BaseModel):
    """Returned immediately after answering — this is where the answer is revealed.

    One shape covers both question types, with the other type's fields null. A
    multiple-choice result carries `selected_index` and `correct_index`; a written
    one carries the verdict, the rubric breakdown and the feedback.
    """

    question_id: uuid.UUID
    question_type: QuestionType
    explanation: str
    source: SourceRef | None

    # Multiple choice.
    selected_index: int | None = None
    correct_index: int | None = None

    # Null for an `uncertain` verdict and for a grading failure: neither is a
    # boolean outcome, and coercing one would report a judgement nobody made.
    is_correct: bool | None = None

    # Short answer.
    response_text: str | None = None
    verdict: AnswerVerdict | None = None
    grading_state: GradingState = GradingState.NOT_REQUIRED
    rubric_results: list[ConceptResultRead] = Field(default_factory=list)
    feedback: str | None = None
    # Shown only after the answer is submitted, exactly like `correct_index`.
    reference_answer: str | None = None
    # True when the answer was recorded but could not be marked, so the UI can say
    # so instead of showing a silent blank.
    grading_failed: bool = False


class AttemptRead(ORMModel):
    id: uuid.UUID
    quiz_id: uuid.UUID
    started_at: datetime
    completed_at: datetime | None
    score_percent: float | None
    correct_count: int


class AttemptSummary(AttemptRead):
    """The completed-attempt view, with per-question results."""

    quiz_title: str
    question_count: int
    results: list[AnswerResult]


# --- Mastery -------------------------------------------------------------------


class TopicMasteryRead(BaseModel):
    topic_id: uuid.UUID
    topic_name: str
    mastery_score: float
    band: str
    band_label: str
    questions_attempted: int
    correct_answers: int
    # None when nothing has been attempted — not zero, which would read as failure.
    accuracy: float | None
    last_practised_at: datetime | None


class CourseMasteryRead(BaseModel):
    course_id: uuid.UUID
    topics: list[TopicMasteryRead]
    # Averaged over attempted topics only; None when nothing has been attempted.
    overall_mastery: float | None
    topics_total: int
    topics_started: int
    topics_strong: int
    questions_answered: int
    correct_answers: int
    accuracy: float | None
    strongest_topic: str | None
    weakest_topic: str | None


class RecommendationRead(BaseModel):
    kind: str
    title: str
    detail: str
    topic_id: uuid.UUID | None
    topic_name: str | None


# --- Flashcards ----------------------------------------------------------------


class FlashcardGenerateRequest(BaseModel):
    topic_ids: list[uuid.UUID] = Field(default_factory=list)
    # When true, ANCHOR picks the topics using the same deterministic priority the
    # adaptive quiz uses.
    weak_topics_only: bool = False


class FlashcardRead(ORMModel):
    id: uuid.UUID
    course_id: uuid.UUID
    topic_id: uuid.UUID
    topic_name: str
    front: str
    back: str
    source: SourceRef | None
    created_at: datetime

File: frontend/src/services/api/learning.ts
import { apiRequest } from '@/services/api/client';
import type { ISODateString } from '@/types/domain';

/* -------------------------------------------------------------------------- */
/*  Wire types                                                                 */
/* -------------------------------------------------------------------------- */

interface TopicDto {
  id: string;
  course_id: string;
  name: string;
  description: string;
  is_active: boolean;
  created_at: ISODateString;
}

interface SourceDto {
  document_id: string;
  document_name: string;
  page_number: number | null;
  chunk_id: string;
}

interface QuizQuestionDto {
  id: string;
  position: number;
  question_text: string;
  question_type: QuestionType;
  /** Null for short answers — nothing to choose from. */
  options: string[] | null;
  difficulty: Difficulty;
  topic_id: string;
  topic_name: string;
}

interface QuizDto {
  id: string;
  course_id: string;
  title: string;
  mode: QuizMode;
  selection_rationale: string;
  difficulty_plan: Record<string, number>;
  question_count: number;
  created_at: ISODateString;
  questions?: QuizQuestionDto[];
}

interface AttemptDto {
  id: string;
  quiz_id: string;
  started_at: ISODateString;
  completed_at: ISODateString | null;
  score_percent: number | null;
  correct_count: number;
}

interface ConceptResultDto {
  concept: string;
  satisfied: boolean;
}

interface AnswerResultDto {
  question_id: string;
  question_type: QuestionType;
  explanation: string;
  source: SourceDto | null;
  selected_index: number | null;
  correct_index: number | null;
  is_correct: boolean | null;
  response_text: string | null;
  verdict: AnswerVerdict | null;
  grading_state: GradingState;
  rubric_results: ConceptResultDto[];
  feedback: string | null;
  reference_answer: string | null;
  grading_failed: boolean;
}

interface AttemptSummaryDto extends AttemptDto {
  quiz_title: string;
  question_count: number;
  results: AnswerResultDto[];
}

interface TopicMasteryDto {
  topic_id: string;
  topic_name: string;
  mastery_score: number;
  band: MasteryBand;
  band_label: string;
  questions_attempted: number;
  correct_answers: number;
  accuracy: number | null;
  last_practised_at: ISODateString | null;
}

interface CourseMasteryDto {
  course_id: string;
  topics: TopicMasteryDto[];
  overall_mastery: number | null;
  topics_total: number;
  topics_started: number;
  topics_strong: number;
  questions_answered: number;
  correct_answers: number;
  accuracy: number | null;
  strongest_topic: string | null;
  weakest_topic: string | null;
}

interface RecommendationDto {
  kind: string;
  title: string;
  detail: string;
  topic_id: string | null;
  topic_name: string | null;
}

interface FlashcardDto {
  id: string;
  course_id: string;
  topic_id: string;
  topic_name: string;
  front: string;
  back: string;
  source: SourceDto | null;
  created_at: ISODateString;
}

/* -------------------------------------------------------------------------- */
/*  Domain types                                                               */
/* -------------------------------------------------------------------------- */

export const DIFFICULTIES = ['easy', 'medium', 'hard'] as const;
export type Difficulty = (typeof DIFFICULTIES)[number];

/**
 * How a quiz's topics were chosen. `exam` is coverage-first under a deadline; the
 * backend decides in every case — the model only writes the questions.
 */
export type QuizMode = 'standard' | 'adaptive' | 'exam';

/** One stored question's format. Mirrors `QuestionType` in the backend models. */
export type QuestionType = 'mcq' | 'short_answer';

/** What a generated quiz should contain. Multiple choice stays the default. */
export const QUIZ_FORMATS = ['mcq', 'short_answer', 'mixed'] as const;
export type QuizFormat = (typeof QUIZ_FORMATS)[number];

/**
 * The outcome of marking one written answer.
 *
 * `uncertain` is a real result, not an error: the grader could not judge the
 * answer, so it neither rewards nor penalises the student and is left out of the
 * score denominator. The UI must show it as unmarked, never as wrong.
 */
export type AnswerVerdict =
  | 'correct'
  | 'partially_correct'
  | 'incorrect'
  | 'uncertain';

/** Whether marking happened at all. `failed` means the provider could not be
 * reached; the answer is stored and mastery is untouched. */
export type GradingState = 'not_required' | 'graded' | 'failed';

/** Mirrors the bands in backend/app/services/learning/mastery.py. */
export type MasteryBand = 'not_started' | 'needs_practice' | 'developing' | 'strong';

export interface Topic {
  id: string;
  courseId: string;
  name: string;
  description: string;
  isActive: boolean;
}

export interface QuizSource {
  documentId: string;
  documentName: string;
  /** Null for formats without real pages (TXT, Markdown) — never a fabricated 1. */
  pageNumber: number | null;
  chunkId: string;
}

export interface QuizQuestion {
  id: string;
  position: number;
  questionText: string;
  questionType: QuestionType;
  /** Null for short answers. The reference answer and rubric are never sent to
   * the client before the student submits. */
  options: string[] | null;
  difficulty: Difficulty;
  topicId: string;
  topicName: string;
}

export interface Quiz {
  id: string;
  courseId: string;
  title: string;
  mode: QuizMode;
  selectionRationale: string;
  difficultyPlan: Record<string, number>;
  questionCount: number;
  createdAt: ISODateString;
  questions: QuizQuestion[];
}

export interface Attempt {
  id: string;
  quizId: string;
  startedAt: ISODateString;
  completedAt: ISODateString | null;
  scorePercent: number | null;
  correctCount: number;
}

export interface ConceptResult {
  concept: string;
  satisfied: boolean;
}

export interface AnswerResult {
  questionId: string;
  questionType: QuestionType;
  explanation: string;
  source: QuizSource | null;

  /** Multiple choice. Null on a written answer. */
  selectedIndex: number | null;
  correctIndex: number | null;
  /** Null for a partial or uncertain verdict — neither is a boolean outcome. */
  isCorrect: boolean | null;

  /** Short answer. */
  responseText: string | null;
  verdict: AnswerVerdict | null;
  gradingState: GradingState;
  rubricResults: ConceptResult[];
  feedback: string | null;
  referenceAnswer: string | null;
  gradingFailed: boolean;
}

export interface AttemptSummary extends Attempt {
  quizTitle: string;
  questionCount: number;
  results: AnswerResult[];
}

export interface TopicMastery {
  topicId: string;
  topicName: string;
  masteryScore: number;
  band: MasteryBand;
  bandLabel: string;
  questionsAttempted: number;
  correctAnswers: number;
  accuracy: number | null;
  lastPractisedAt: ISODateString | null;
}

export interface CourseMastery {
  courseId: string;
  topics: TopicMastery[];
  overallMastery: number | null;
  topicsTotal: number;
  topicsStarted: number;
  topicsStrong: number;
  questionsAnswered: number;
  correctAnswers: number;
  accuracy: number | null;
  strongestTopic: string | null;
  weakestTopic: string | null;
}

export interface Recommendation {
  kind: string;
  title: string;
  detail: string;
  topicId: string | null;
  topicName: string | null;
}

export interface Flashcard {
  id: string;
  courseId: string;
  topicId: string;
  topicName: string;
  front: string;
  back: string;
  source: QuizSource | null;
}

/* -------------------------------------------------------------------------- */
/*  Mapping                                                                    */
/* -------------------------------------------------------------------------- */

function toSource(dto: SourceDto | null): QuizSource | null {
  if (!dto) return null;
  return {
    documentId: dto.document_id,
    documentName: dto.document_name,
    pageNumber: dto.page_number,
    chunkId: dto.chunk_id,
  };
}

function toTopic(dto: TopicDto): Topic {
  return {
    id: dto.id,
    courseId: dto.course_id,
    name: dto.name,
    description: dto.description,
    isActive: dto.is_active,
  };
}

function toQuestion(dto: QuizQuestionDto): QuizQuestion {
  return {
    id: dto.id,
    position: dto.position,
    questionText: dto.question_text,
    questionType: dto.question_type,
    options: dto.options,
    difficulty: dto.difficulty,
    topicId: dto.topic_id,
    topicName: dto.topic_name,
  };
}

function toQuiz(dto: QuizDto): Quiz {
  return {
    id: dto.id,
    courseId: dto.course_id,
    title: dto.title,
    mode: dto.mode,
    selectionRationale: dto.selection_rationale,
    difficultyPlan: dto.difficulty_plan,
    questionCount: dto.question_count,
    createdAt: dto.created_at,
    questions: (dto.questions ?? []).map(toQuestion),
  };
}

function toAttempt(dto: AttemptDto): Attempt {
  return {
    id: dto.id,
    quizId: dto.quiz_id,
    startedAt: dto.started_at,
    completedAt: dto.completed_at,
    scorePercent: dto.score_percent,
    correctCount: dto.correct_count,
  };
}

function toAnswerResult(dto: AnswerResultDto): AnswerResult {
  return {
    questionId: dto.question_id,
    questionType: dto.question_type,
    explanation: dto.explanation,
    source: toSource(dto.source),
    selectedIndex: dto.selected_index,
    correctIndex: dto.correct_index,
    isCorrect: dto.is_correct,
    responseText: dto.response_text,
    verdict: dto.verdict,
    gradingState: dto.grading_state,
    rubricResults: dto.rubric_results.map((row) => ({
      concept: row.concept,
      satisfied: row.satisfied,
    })),
    feedback: dto.feedback,
    referenceAnswer: dto.reference_answer,
    gradingFailed: dto.grading_failed,
  };
}

function toMastery(dto: TopicMasteryDto): TopicMastery {
  return {
    topicId: dto.topic_id,
    topicName: dto.topic_name,
    masteryScore: dto.mastery_score,
    band: dto.band,
    bandLabel: dto.band_label,
    questionsAttempted: dto.questions_attempted,
    correctAnswers: dto.correct_answers,
    accuracy: dto.accuracy,
    lastPractisedAt: dto.last_practised_at,
  };
}

function toFlashcard(dto: FlashcardDto): Flashcard {
  return {
    id: dto.id,
    courseId: dto.course_id,
    topicId: dto.topic_id,
    topicName: dto.topic_name,
    front: dto.front,
    back: dto.back,
    source: toSource(dto.source),
  };
}

/* -------------------------------------------------------------------------- */
/*  Endpoints                                                                  */
/* -------------------------------------------------------------------------- */

export async function fetchTopics(courseId: string): Promise<Topic[]> {
  const dtos = await apiRequest<TopicDto[]>(`/v1/courses/${courseId}/topics`);
  return dtos.map(toTopic);
}

export interface TopicSyncStatus {
  courseId: string;
  /**
   * False only when the course has processed material the topic set was not
   * derived from. A course with nothing processed reports true — there is
   * nothing to extract, so prompting the student would be noise.
   */
  topicsAreCurrent: boolean;
  topicCount: number;
  readyDocumentCount: number;
}

interface TopicSyncStatusDto {
  course_id: string;
  topics_are_current: boolean;
  topic_count: number;
  ready_document_count: number;
}

/** Deterministic and cheap — no model call, safe on every page load. */
export async function fetchTopicSyncStatus(courseId: string): Promise<TopicSyncStatus> {
  const dto = await apiRequest<TopicSyncStatusDto>(
    `/v1/courses/${courseId}/topics/status`,
  );
  return {
    courseId: dto.course_id,
    topicsAreCurrent: dto.topics_are_current,
    topicCount: dto.topic_count,
    readyDocumentCount: dto.ready_document_count,
  };
}

export interface TopicExtraction {
  created: Topic[];
  reactivated: Topic[];
  deactivated: Topic[];
  unchanged: Topic[];
}

export async function extractTopics(courseId: string): Promise<TopicExtraction> {
  const dto = await apiRequest<{
    created: TopicDto[];
    reactivated: TopicDto[];
    deactivated: TopicDto[];
    unchanged: TopicDto[];
  }>(`/v1/courses/${courseId}/topics/extract`, { method: 'POST' });

  return {
    created: dto.created.map(toTopic),
    reactivated: dto.reactivated.map(toTopic),
    deactivated: dto.deactivated.map(toTopic),
    unchanged: dto.unchanged.map(toTopic),
  };
}

export interface GenerateQuizInput {
  mode: QuizMode;
  questionCount: number;
  topicIds?: string[];
  difficulty?: Difficulty | null;
  quizFormat?: QuizFormat;
}

export async function generateQuiz(
  courseId: string,
  input: GenerateQuizInput,
): Promise<Quiz> {
  const dto = await apiRequest<QuizDto>(`/v1/courses/${courseId}/quizzes`, {
    method: 'POST',
    body: {
      mode: input.mode,
      question_count: input.questionCount,
      topic_ids: input.topicIds ?? [],
      difficulty: input.difficulty ?? null,
    },
  });
  return toQuiz(dto);
}

export async function fetchQuizzes(courseId: string): Promise<Quiz[]> {
  const dtos = await apiRequest<QuizDto[]>(`/v1/courses/${courseId}/quizzes`);
  return dtos.map(toQuiz);
}

export async function fetchQuiz(quizId: string): Promise<Quiz> {
  return toQuiz(await apiRequest<QuizDto>(`/v1/quizzes/${quizId}`));
}

export async function startAttempt(quizId: string): Promise<Attempt> {
  return toAttempt(
    await apiRequest<AttemptDto>(`/v1/quizzes/${quizId}/attempts`, { method: 'POST' }),
  );
}

export async function submitAnswer(
  attemptId: string,
  questionId: string,
  selectedIndex: number,
  answeredInSeconds?: number,
): Promise<AnswerResult> {
  return toAnswerResult(
    await apiRequest<AnswerResultDto>(`/v1/attempts/${attemptId}/answers`, {
      method: 'POST',
      body: {
        question_id: questionId,
        selected_index: selectedIndex,
        ...(answeredInSeconds === undefined
          ? {}
          : { answered_in_seconds: answeredInSeconds }),
      },
    }),
  );
}

/**
 * Submit one written answer.
 *
 * Slower than `submitAnswer`: marking happens inline, because the student is
 * waiting for feedback and a queued verdict would mean showing them an answer
 * with no result attached.
 */
export async function submitShortAnswer(
  attemptId: string,
  questionId: string,
  responseText: string,
  answeredInSeconds?: number,
): Promise<AnswerResult> {
  return toAnswerResult(
    await apiRequest<AnswerResultDto>(`/v1/attempts/${attemptId}/short-answers`, {
      method: 'POST',
      body: {
        question_id: questionId,
        response_text: responseText,
        ...(answeredInSeconds === undefined
          ? {}
          : { answered_in_seconds: answeredInSeconds }),
      },
    }),
  );
}

export async function completeAttempt(attemptId: string): Promise<AttemptSummary> {
  const dto = await apiRequest<AttemptSummaryDto>(
    `/v1/attempts/${attemptId}/complete`,
    { method: 'POST' },
  );
  return {
    ...toAttempt(dto),
    quizTitle: dto.quiz_title,
    questionCount: dto.question_count,
    results: dto.results.map(toAnswerResult),
  };
}

export async function fetchMastery(courseId: string): Promise<CourseMastery> {
  const dto = await apiRequest<CourseMasteryDto>(`/v1/courses/${courseId}/mastery`);
  return {
    courseId: dto.course_id,
    topics: dto.topics.map(toMastery),
    overallMastery: dto.overall_mastery,
    topicsTotal: dto.topics_total,
    topicsStarted: dto.topics_started,
    topicsStrong: dto.topics_strong,
    questionsAnswered: dto.questions_answered,
    correctAnswers: dto.correct_answers,
    accuracy: dto.accuracy,
    strongestTopic: dto.strongest_topic,
    weakestTopic: dto.weakest_topic,
  };
}

export async function fetchRecommendations(courseId: string): Promise<Recommendation[]> {
  const dtos = await apiRequest<RecommendationDto[]>(
    `/v1/courses/${courseId}/recommendations`,
  );
  return dtos.map((dto) => ({
    kind: dto.kind,
    title: dto.title,
    detail: dto.detail,
    topicId: dto.topic_id,
    topicName: dto.topic_name,
  }));
}

export async function fetchAttempts(courseId: string): Promise<Attempt[]> {
  const dtos = await apiRequest<AttemptDto[]>(`/v1/courses/${courseId}/attempts`);
  return dtos.map(toAttempt);
}

export async function fetchFlashcards(courseId: string): Promise<Flashcard[]> {
  const dtos = await apiRequest<FlashcardDto[]>(`/v1/courses/${courseId}/flashcards`);
  return dtos.map(toFlashcard);
}

export interface GenerateFlashcardsInput {
  topicIds?: string[];
  weakTopicsOnly?: boolean;
}

export async function generateFlashcards(
  courseId: string,
  input: GenerateFlashcardsInput,
): Promise<Flashcard[]> {
  const dtos = await apiRequest<FlashcardDto[]>(`/v1/courses/${courseId}/flashcards`, {
    method: 'POST',
    body: {
      topic_ids: input.topicIds ?? [],
      weak_topics_only: input.weakTopicsOnly ?? false,
    },
  });
  return dtos.map(toFlashcard);
}
